src/software/ai/evaluation/time_to_travel.cpp

```diff
@@ -0,0 +1,70 @@
+#include <cmath>
+
+#include "software/ai/evaluation/time_to_travel.h"
+
+Duration getTimeToTravelDistance(const double distance, const double max_velocity,
+                                 const double max_acceleration,
+                                 const double initial_velocity,
+                                 const double final_velocity)
+{
+    // Bound all values to be realistic
+    double d_total = std::max(0.0, distance);
+    double v_max   = std::max(0.0, max_velocity);
+    double v_i     = std::clamp(initial_velocity, -max_velocity, max_velocity);
+    double v_f     = std::clamp(final_velocity, 0.0, max_velocity);
+    double a_max   = std::max(1e-6, max_acceleration);
+
+    // Minimum distance required to accelerate/decelerate from initial to final velocity
+    double dist_required_to_reach_v_f =
+        std::abs(std::pow(v_f, 2) - std::pow(v_i, 2)) / (2 * a_max);
+    if (dist_required_to_reach_v_f > d_total)
+    {
+        // Accelerating/decelerating instantly from initial to final velocity is not
+        // possible with in the given distance, given the robot's max acceleration,
+        // therefore the robot can not reach the desired final velocity. Calculate how
+        // long it will take for the robot to accelerate towards final velocity over
+        // distance
+        double a_max_signed = a_max;
+        if (v_f < v_i)
+        {
+            // Robot must decelerate, so it has negative acceleration
+            a_max_signed *= -1.0;
+        }
+        // t = (-Vi + sqrt(Vi^2 + 2 * a * d)) / a
+        double t_total =
+            (-v_i + std::sqrt(std::pow(v_i, 2) + 2 * a_max_signed * d_total)) /
+            a_max_signed;
+        return Duration::fromSeconds(t_total);
+    }
+
+    // Total travel time if the robot is always accelerating at max acceleration (i.e.
+    // accelerating to the highest value possible given the distance, and then
+    // decelerating to final velocity)
+    double t_total =
```

Comment:
curious what the derivation here is. i thought kinematic equations only applied when there is constant acceleration, but my understanding from the comment above is that you are accelerating and then decelerating?

Reply:
Yeah you're correct. I wasn't sure how exactly I should go about explaining this. 
The formula is derived from combining regular 1D kinematics equations under the assumptions that the robot will travel a distance as fast as possible (i.e. Accelerating as much as possible, and decelerating as late as possible to reach the desired final velocity). Under those assumptions I derived the formula for the total time. 
To help visualize it, Ive made this [desmos](https://www.desmos.com/calculator/jz9txhttq9) link with a velocity vs time graph, where the area under the graph should be equal to the distance we want to travel.

Reply:
this was very helpful. i was also afraid the equations wouldnt behave well near `a=1e-6` but i can see now that the limits work out